src/Mutation.tsx

```diff
@@ -0,0 +1,93 @@
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import ApolloClient from 'apollo-client';
+const invariant = require('invariant');
+import { DocumentNode } from 'graphql';
+
+import { OperationVariables } from './types';
+
+// refetchQueries?: string[] | PureQueryOptions[];
+// update?: MutationUpdaterFn;
+// notifyOnNetworkStatusChange?: boolean;
+
+export interface MutationProps {
+  mutation: DocumentNode;
+  variables?: OperationVariables;
+  optimisticResponse?: Object;
+  children: (mutateFn: () => void, result?: any) => React.ReactNode;
+}
+
+export interface MutationState {
+  notCalled: boolean;
+  error?: string;
+  data?: any;
+  loading?: boolean;
+}
+
+class Mutation extends React.Component<MutationProps, MutationState> {
+  private client: ApolloClient<any>;
+
+  static contextTypes = {
+    client: PropTypes.object.isRequired,
+  };
+
+  constructor(props: any, context: any) {
+    super(props, context);
+
+    invariant(
+      !!context.client,
+      `Could not find "client" in the context of Mutation. Wrap the root component in an <ApolloProvider>`,
+    );
+    this.client = context.client;
+
+    this.state = {
+      notCalled: true,
+    };
+  }
+
+  private runMutation = async () => {
+    const { mutation, variables, optimisticResponse } = this.props;
+
+    this.setState({
+      loading: true,
+      error: undefined,
+      data: undefined,
+      notCalled: false,
+    });
+
+    try {
+      const response = await this.client.mutate({
```

Comment:
This will ultimately need parameterized types but https://github.com/apollographql/apollo-client/issues/2795 needs fixed first to use them.